AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java

```diff
@@ -0,0 +1,166 @@
+    assertEquals(expiration, message.getExpirationTimestamp());
+    assertTrue(message.isRead());
+    assertNull(message.getData());
+
+    assertNull(message.getImageFilePath());
+    assertNull(message.getImageUrl());
+  }
+
+  @Test
+  public void testInvalidMessageIdIsRejected() {
+    Date delivery = new Date(100);
+    Date expiration = new Date(200);
+    HashMap<String, Object> map = new HashMap<>();
+    map.put(Constants.Keys.MESSAGE_DATA, new HashMap<String, Object>());
+    map.put(Constants.Keys.DELIVERY_TIMESTAMP, delivery.getTime());
+    map.put(Constants.Keys.EXPIRATION_TIMESTAMP, expiration.getTime());
+    map.put(Constants.Keys.IS_READ, true);
+
+    LeanplumInboxMessage invalidMessage = LeanplumInboxMessage.createFromJsonMap("messageId", map);
+    assertNull(invalidMessage);
+  }
+
+  @Test
+  public void testReadAndUpdateMessageCount() {
+    Date delivery = new Date(100);
+    Date expiration = new Date(200);
+    HashMap<String, Object> map = new HashMap<>();
+    map.put(Constants.Keys.MESSAGE_DATA, new HashMap<String, Object>());
+    map.put(Constants.Keys.DELIVERY_TIMESTAMP, delivery.getTime());
+    map.put(Constants.Keys.EXPIRATION_TIMESTAMP, expiration.getTime());
+    map.put(Constants.Keys.IS_READ, false);
+    LeanplumInboxMessage message = LeanplumInboxMessage
+        .createFromJsonMap("messageId##00", map);
+    int intialUnreadCount = LeanplumInbox.getInstance().unreadCount();
+
+    assertEquals(false, message.isRead());
+
+    message.read();
+
+    assertEquals(true, message.isRead());
+    assertEquals(intialUnreadCount-1, LeanplumInbox.getInstance().unreadCount());
+  }
+
+  @Test
+  public void testImageURL() {
+    LeanplumInbox.getInstance().disableImagePrefetching();
+    ResponseHelper.seedResponse("/responses/newsfeed_response.json");
+    LeanplumInbox.getInstance().downloadMessages();
+    List<LeanplumInboxMessage> messagesList = LeanplumInbox.getInstance().allMessages();
+    LeanplumInboxMessage imageMessage = messagesList.get(0);
+
+    String actualUrl = imageMessage.getImageUrl().toString();
+
+    assertEquals("http://bit.ly/2GzJxxx", actualUrl);
+  }
+
+  @Test
+  public void testImageFilepathIsReturnedIfPrefetchingEnabled() {
+    ResponseHelper.seedResponse("/responses/newsfeed_response.json");
+    LeanplumInbox.getInstance().downloadMessages();
+    LeanplumInboxMessage imageMessage = LeanplumInbox.getInstance().allMessages().get(0);
+
+    String imageFilePath = imageMessage.getImageFilePath();
+
+    assertNotNull(imageFilePath);
+  }
+
+  @Test
+  public void testOpenAction() {
+    ResponseHelper.seedResponse("/responses/newsfeed_response.json");
+    LeanplumInbox.getInstance().downloadMessages();
+    LeanplumInboxMessage imageMessage = LeanplumInbox.getInstance().allMessages().get(1);
+    imageMessage.read();
+
+    HashMap actionName = imageMessage.getContext().objectNamed("Open action");
+
+    assertEquals("Alert", actionName.get("__name__"));
```

Comment:
There you checked just that you have it in ActionContext. You need also check that when you read this message - the open action was run.

Reply:
The test for running the open action should be in ActionContext since it is the responsibility of ActionContext to display the message. LeanplumInboxMessage only puts messages into ActionContext, therefore the unit test should test upto the behavior of the class. I checked in ActionContext and there is a test that covers open actions in ActionContext that milos wrote.

Reply:
yes, I understand this. In other words, you need to verify that  runTrackedActionNamed was called when you call 
https://github.com/Leanplum/Leanplum-Android-SDK/pull/164/files/200d38a529c046090da0e1435ba488a452be2a53#diff-8e13c25a7be11e2043a2f72ba823c3f6R138

Reply:
https://github.com/Leanplum/Leanplum-Android-SDK/blob/master/AndroidSDKCore/src/main/java/com/leanplum/LeanplumInboxMessage.java#L192

Reply:
@Enuviel yes agreed, like I mentioned earlier the read method calls runTrackedActionNamed(), the functionality of that method and executing of an open action in general is the responsibility of the ActionContext class, not of LeanplumInboxMessage class. The responsibility of LeanplumInboxMessage is to correctly form an message entity and interact with the LeanplumInbox. In my test above I'm testing to ensure that the message entity has the correct open action, in this case "Alert", and testing to make sure that it can be correctly accessed by ActionContext. I don't disagree that testing whether ActionContext properly executes an open action is a good thing, however I'm saying that test should be in ActionContextTest not in LeanplumInboxMessageTest